### PR TITLE
Spacer integration

### PR DESCRIPTION
**Pull request by beijbom**
This addresses most of https://github.com/beijbom/coralnet/issues/264.

# Summary
I wanted to try not making this PR too large. But I'm afraid I failed. Here are the main changes:
- Use the new spacer in `vision_backend/tasks.py` and `task_helpers.py`.
- Use a swappable queue to allow `SQSQueue` for pushing jobs to AWS and `LocalQueue` for processing jobs locally. The latter is for unit-tests and reg-tests. Note that the only thing we change here is the Queue. The actual processing is done by `pyspacer` in both cases.
- PEP8 compliance in the files I touched. I didn't go over all files in the app.
- Refreshed backend regression tests.
- Deleted `vision_backend/scrips.py`.

# How to test
All unit-test should pass. Run regression tests by `python manage.py vb_regtests`. There are several ways to run the tests: check out `python manage.py vb_regtests --help` for details.
## Local eager
This is the simplest setting. Local backend, no celery needed, local queue.
```
CELERY_ALWAYS_EAGER = True
SPACER_QUEUE_CHOICE = 'vision_backend.queues.LocalQueue'
from .storage_local import *
```

## Local celery
Here everything is still local, but jobs are executed asynchronously using celery. Make sure the celery worker is running before starting the test: `celery -A config worker`.
```
CELERY_ALWAYS_EAGER = False
SPACER_QUEUE_CHOICE = 'vision_backend.queues.LocalQueue'
from .storage_local import *
```

## S3 celery
Still using the LocalQueue, but storage is now on S3. 
```
CELERY_ALWAYS_EAGER = False
SPACER_QUEUE_CHOICE = 'vision_backend.queues.LocalQueue'
from .storage_s3 import *.
```

## AWS ECS cluster. 
This uses the AWS SQS and ECS to process the jobs. The jobs are submitted to a test_queue, so this is safe to run.
```
CELERY_ALWAYS_EAGER = False
SPACER_QUEUE_CHOICE = 'vision_backend.queues.SQSQueue'
from .storage_s3 import *
```
Etc. etc. Most combinations work. The only requirement is that if you use the `SQSQueue` you need to use the `storage_s3` setting. Also, if you use `.storage_local` you need to add these lines to your personal dev_settings to allow access to fixtures and spacer models from s3.
```
AWS_ACCESS_KEY_ID = get_secret('AWS_ACCESS_KEY_ID')
AWS_SECRET_ACCESS_KEY = get_secret('AWS_SECRET_ACCESS_KEY')
os.environ['SPACER_AWS_ACCESS_KEY_ID'] = AWS_ACCESS_KEY_ID
os.environ['SPACER_AWS_SECRET_ACCESS_KEY'] = AWS_SECRET_ACCESS_KEY
```

# Notes
- I removed `FORCE_NO_BACKEND_SUBMIT` since now we can always submit jobs now, we just need to make sure its the right queue.
- I added a `FORCE_DUMMY_EXTRACTOR` to be set for tests. This forces tasks to use the spacer `dummy` extractor, which is very fast.
- I added a new secret: `SPACER_JOB_HASH`. This is a unique string that ensures we only collect jobs for the correct server.
- I added configs for the two sets of SQS I created.
- I had to mute the logger during testing. I think it has to do with spacer using the logging module.
- I tried enabling `SuccessTest.test_done()` in `vision_backend_api/tests/test_deploy.py`. To do that I needed to actually train a classifier, which we now can do quite quickly. So I used a similar design as in `TrainClassifierTest` in vision_backend/tests/test_tasks.py. But for some reason, it didn't work here. I suspect is has to do with the test base classes. 
- The code in `regtests_utils.py` is horrible but I couldn't be bothered fixing it. I wasn't quite sure how and didn't feel like going down that hole.

- [x] Merged in the swappable backend from https://github.com/beijbom/coralnet/pull/265.
- [x] Port extract_features.
- [x] Switch to local spacer dev-version.
- [x] Port train_classifier.
- [x] Write job_token parser, and fix previous classifier updates.
- [x] Port classify_features.
- [x] Port classify_image (aka deploy).
- [x] Update deploy parse test.
- [x] Make all unit-tests pass.
- [x] Fix `add_scores` and `add_annotations`.
- [x] Add spacer_job_hash.
- [x] Rename `MockBackend` to `MockQueue`, etc.
- [x] Remove FORCE_NO_BACKEND_SUBMIT?
- [x] Change from print to log in spacer.
- [x] Make sure new `valres` is parsed correctly.
- [x] Run reg-tests using MockQueue with both storage classes.
- [x] Write ValRes size tests to catch previous bug.
- [x] Make backwards compatibility tests for ValRes.
- [x] Make new spacer release.
- [x] Switch to new spacer release.
- [x] Update reg-tests management command.
- [x] Clean up scripts.py in vision_backend.
- [x] Update spacer ECS with latest docker.
- [x] Run reg-tests using SQSQueue with S3 storage.

This PR has two minor "friend-PRs" on the spacer side. These are already merged to master and included in the `0.2.6` release.
https://github.com/beijbom/pyspacer/pull/25
https://github.com/beijbom/pyspacer/pull/26

### CODE REVIEW

**Comment by beijbom** [comment]
One thing to note is that `celery beat` exited with some decoding error. `celery worker` seemed to run fine however. 
```
(coralnet3) oscarbook2:project beijbom$ celery -A config beat
celery beat v3.1.23 (Cipater) is starting.
__    -    ... __   -        _
Configuration ->
    . broker -> redis://localhost:6379//
    . loader -> celery.loaders.app.AppLoader
    . scheduler -> celery.beat.PersistentScheduler
    . db -> celerybeat-schedule
    . logfile -> [stderr]@%INFO
    . maxinterval -> now (0s)
[2020-06-05 23:42:04,641: INFO/MainProcess] beat: Starting...
[2020-06-05 23:42:04,656: CRITICAL/MainProcess] beat raised exception <class 'UnicodeDecodeError'>: UnicodeDecodeError('ascii', b'\x07\xe4\x04\n\x0f&2\x075\xf4', 1, 2, 'ordinal not in range(128)')
Traceback (most recent call last):
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/kombu/utils/__init__.py", line 323, in __get__
    return obj.__dict__[self.__name__]
KeyError: 'scheduler'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/shelve.py", line 111, in __getitem__
    value = self.cache[key]
KeyError: 'entries'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/apps/beat.py", line 112, in start_scheduler
    beat.start()
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/beat.py", line 470, in start
    humanize_seconds(self.scheduler.max_interval))
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/kombu/utils/__init__.py", line 325, in __get__
    value = obj.__dict__[self.__name__] = self.__get(obj)
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/beat.py", line 512, in scheduler
    return self.get_scheduler()
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/beat.py", line 507, in get_scheduler
    lazy=lazy)
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/utils/imports.py", line 53, in instantiate
    return symbol_by_name(name)(*args, **kwargs)
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/beat.py", line 358, in __init__
    Scheduler.__init__(self, *args, **kwargs)
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/beat.py", line 185, in __init__
    self.setup_schedule()
  File "/Users/beijbom/.virtualenvs/coralnet3/lib/python3.6/site-packages/celery/beat.py", line 388, in setup_schedule
    self._store['entries']
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/shelve.py", line 114, in __getitem__
    value = Unpickler(f).load()
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe4 in position 1: ordinal not in range(128)
```

**Comment by beijbom** [comment]
This tests fails intermittently. I wasn't able to find a real pattern: 
```
======================================================================
FAIL: test_all_modes (images.tests.test_commands.RemovePointOutliersTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/beijbom/coralnet/project/images/tests/test_commands.py", line 87, in test_all_modes
    self.assertIn("{} 4".format(img2.pk), stdout_text)
AssertionError: '102 4' not found in '102 3\n103 4'
```

**Comment by beijbom** [comment]
@StephenChan : this is finally ready to review. Sorry, it ended up being quite large -- it was just a lot of things that were densely connected and needed to be done at the same time.

I suggest you focus on looking at the added SETTINGS and the design of the `LocalQueue` vs `SQSQueue`. If you think this looks OK, I'd suggest we merge this to the base-branch, and then perhaps you take a pass and write some more tests on the vision_backend module? 

I can then do another round of cleanup / refactoring once we have better test coverage. I was able to "unskip" some of the test you had written in `test_tasks.py` and I added a few more.

Also -- sorry about the long commit history. I'll squash when merging.

**Comment by beijbom** [comment]
Another thing I noticed is when I ran `python manage.py migrate` I got this:
```
(coralnet3) oscarbook2:project beijbom$ python manage.py migrate
Operations to perform:
  Apply all migrations: accounts, admin, andablog, annotations, api_core, auth, authtoken, blog, bug_reporting, contenttypes, easy_thumbnails, errorlogs, flatpages, flatpages_custom, guardian, images, labels, lib, newsfeed, reversion, sessions, sites, taggit, vision_backend
Running migrations:
  No migrations to apply.
  Your models have changes that are not yet reflected in a migration, and so won't be applied.
  Run 'manage.py makemigrations' to make new migrations, and then re-run 'manage.py migrate' to apply them.
```
I did try `manage.py makemigrations` and it crated a bunch of migrations. I didn't commit or run them, but perhaps you want to take a look?

**Comment by beijbom** [comment]
@StephenChan : I pushed a few changes this morning. Nothing major -- I just made the reg-tests more robust to the various configurations.

File: project/lib/storage_backends.py [hunk 1]
```diff
@@ -268,6 +276,17 @@ def path_join(*args):
         # For local storage, we join paths depending on the OS rules.
         return os.path.join(*args)
 
+    def spacer_data_loc(self, key) -> DataLocation:
+        """ Returns a spacer DataLocation object. """
+        return DataLocation(storage_type='filesystem',
+                            key=self.path(key))
+
+    def exists_full(self, full_key) -> bool:
```
Comment (StephenChan) [inline project/lib/storage_backends.py:284]: Judging from commit 86f4ccbbf9c2ee34e277ffab4ea1704a8872b352, it looks like you meant to remove this `exists_full()` method too. Is that right?

File: project/vision_backend/management/commands/vb_regtests.py [hunk 2]
```diff
@@ -1,40 +1,179 @@
...
+        together with `from .storage_s3 import *` fails. 
+        This is likely due to the feature extraction jobs being 
+        submitted before the images are done uploading. 
+        Use CELERY_ALWAYS_EAGER = True if this is a problem.
+        '''
+
+    def create_parser(self, *args, **kwargs):
+        """ This makes the help text more nicely formatted. """
+        parser = super(Command, self).create_parser(*args, **kwargs)
+        parser.formatter_class = RawTextHelpFormatter
```
Comment (StephenChan) [inline project/vision_backend/management/commands/vb_regtests.py:86]: Nice find, I'll have to keep this feature in mind.

File: project/vision_backend/queues.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+import json
+import logging
+import posixpath
+import time
+
+import boto.sqs
+from django.conf import settings
+from django.core.files.storage import get_storage_class
+from django.utils.module_loading import import_string
+from six import StringIO
```
Comment (StephenChan) [inline project/vision_backend/queues.py:11]: Just so you know, you can use `from io import StringIO` instead. That actually goes for Python 2.7 as well as 3.x. If this code works though, then we can leave it for now. I'll clean up all the `six` imports together at some point, anyway.

File: project/vision_backend/queues.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        and return it. If no result is available, return None.
+        """
+        storage = get_storage_class()()
+        dir_names, filenames = storage.listdir('backend_job_res')
+
+        if len(filenames) == 0:
+            return None
+
+        # Sort by filename, which should also put them in job order
+        # because the filenames have timestamps (to second precision)
```
Comment (StephenChan) [inline project/vision_backend/queues.py:132]: I guess you mean microsecond precision here?

File: requirements/local.txt [hunk 5]
```diff
@@ -11,3 +11,8 @@ coverage
 # The version shouldn't really matter here, as it's not part of the running
 # Django project.
 Sphinx
+
+# BeautifulSoup's encoding detection becomes more accurate if chardet or
+# cchardet is installed. chardet is older; cchardet is much faster.
+# Use chardet here since installing cchardet caused trouble on mac.
```
Comment (StephenChan) [inline requirements/local.txt:17]: Hmm, OK. [This cchardet issue comment](https://github.com/PyYoshi/cChardet/issues/37#issuecomment-569967562) suggests that upgrading to Python 3.7 may fix cchardet installation on Mac. Not our immediate priority I guess, but we'll see when we can upgrade from 3.6 to 3.7, then we can look at cchardet again.

File: project/lib/storage_backends.py (hunk 1 above)
Comment (beijbom) [reply project/lib/storage_backends.py:284]: Yes. thanks for catching that.

File: project/vision_backend/queues.py (hunk 3 above)
Comment (beijbom) [reply project/vision_backend/queues.py:11]: I see. Yeah, this import was a left-over from your "swappable backend" branch. I'll just leave it for now and we can purge along with the other when we deprecate 2.7.

File: requirements/local.txt (hunk 5 above)
Comment (beijbom) [reply requirements/local.txt:17]: sounds good.

File: project/vision_backend/queues.py (hunk 4 above)
Comment (beijbom) [reply project/vision_backend/queues.py]: Yes. Thanks. Fixed.

File: project/vision_backend/queues.py (hunk 3 above)
Comment (StephenChan) [reply project/vision_backend/queues.py:11]: Ah, I see. Right, I didn't understand all the Python 2 vs. 3 imports as well when I made that branch.

File: project/vision_backend/management/commands/vb_regtests.py [hunk 6]
```diff
@@ -1,40 +1,179 @@
...
+        os.environ['SPACER_AWS_ACCESS_KEY_ID'] = AWS_ACCESS_KEY_ID
+        os.environ['SPACER_AWS_SECRET_ACCESS_KEY'] = AWS_SECRET_ACCESS_KEY
+        ```
+                
+        NOTE2: 
+        Sometimes using `CELERY_ALWAYS_EAGER = True`
+        together with `from .storage_s3 import *` fails. 
+        This is likely due to the feature extraction jobs being 
+        submitted before the images are done uploading. 
+        Use CELERY_ALWAYS_EAGER = True if this is a problem.
```
Comment (StephenChan) [inline project/vision_backend/management/commands/vb_regtests.py:80]: I guess this line should say False, not True?

**Comment by StephenChan** [comment]
> this is finally ready to review. Sorry, it ended up being quite large -- it was just a lot of things that were densely connected and needed to be done at the same time.

It did seem like that kind of task to me, so no worries. Exciting to see the new spacer getting put to work in `vision_backend` finally.

> I suggest you focus on looking at the added SETTINGS and the design of the `LocalQueue` vs `SQSQueue`.

Besides the minor comments I made, looks generally good to me.

> I had to mute the logger during testing. I think it has to do with spacer using the logging module.

Hmm, okay. I'll make a note of it in issue #307 in case that could be related.

> I tried enabling `SuccessTest.test_done()` in `vision_backend_api/tests/test_deploy.py`. To do that I needed to actually train a classifier, which we now can do quite quickly. So I used a similar design as in TrainClassifierTest in vision_backend/tests/test_tasks.py. But for some reason, it didn't work here. I suspect is has to do with the test base classes.

Okay, I'll look into this in more detail then.

> The code in regtests_utils.py is horrible but I couldn't be bothered fixing it. I wasn't quite sure how and didn't feel like going down that hole.

Understandable. I can try fixing it later if I get some good ideas on how.

> If you think this looks OK, I'd suggest we merge this to the base-branch, and then perhaps you take a pass and write some more tests on the vision_backend module?
>
> I can then do another round of cleanup / refactoring once we have better test coverage. I was able to "unskip" some of the test you had written in `test_tasks.py` and I added a few more.

Sounds good to me. I'll take care of the zoom/rotation damage assessment first, and perhaps an Edit Metadata quick fix as well, but afterward I'll get on this.

> I did try manage.py makemigrations and it crated a bunch of migrations. I didn't commit or run them, but perhaps you want to take a look?

Ah, looks like most of these are about default string values and help text strings defined in the `models.py` files. They used to be byte strings, now they're Unicode strings since we're on Python 3. That's kind of annoying. I'm not sure whether it's better to take care of the migrations before or after the Beta 2 rollout (or if it doesn't matter). But we can skip them for this PR, and I'll look into it.


As for testing on my end: no unit test failures, and the small-size regtest was able to complete successfully on the 4 example configs you gave in the help text: True local local, False local local, False local S3, False SQS S3.

Though, the 3rd regtest config (False local S3) did encounter a couple of intermittent errors. My first try got `ValueError: Not enough classes to do training (only 1)`, but eventually it tried training again and was successful. Second and third tries stopped short with an error like `OSError: File does not exist: media/backend_job_res/1592365319.3996203.json`. 4th time had no errors. That 'NOTE2' comment looks like you're aware of some intermittent issue with S3, but not sure if it's the same as the ones I got.

Overall though, PR looks good to me. Particularly happy to see spacer can run without S3 storage or SQS.

**Comment by StephenChan** [comment]
I took a more thorough look at the migrations from `makemigrations`. As far as I can tell, they are all just converting byte string literals (of help text, verbose names, choices, etc.) to Unicode string literals. And none of those string literals have non-ASCII characters, from what I noticed. So I don't think the conversion will pose any issues.

Also, the unit tests all pass on Python 3 with the migrations present (the migrations run during the test-runner setup phase). So I think we're good to add the migrations to `beta2-rollout` now. I'll go ahead and commit them to this PR.

**Comment by StephenChan** [comment]
I think my previous comment still stands, except the 'commit to this PR' idea doesn't seem so smart anymore. Oh well. Sorry about the extra clutter.

`django-migration-testcase` isn't a fan of those migrations for some reason - and only on Travis, not my dev machine. Not quite sure why. I don't think it's a big deal though. The test tests the images 0023 migration, and we can remove that migration sometime after the rollout anyway (when handling issue #146). I just skipped that test for now. Just in case, I'll set up a Python 3 environment on our staging server and see if that encounters any issues with the new migrations.

The second skip I just did was an unrelated test which has been unreliable on my Windows dev machine. Sometimes it fails, then the problem goes away later. Looks like the unreliability is passing onto Travis as well.

**Comment by beijbom** [comment]
> I think my previous comment still stands, except the 'commit to this PR' idea doesn't seem so smart anymore. Oh well. Sorry about the extra clutter.

Haha, nah, that seemed like an excellent candidate for a new PR. Owell, doesn't really matter. I'll take a final look now and merge to the feature branch.

> 
> `django-migration-testcase` isn't a fan of those migrations for some reason - and only on Travis, not my dev machine. Not quite sure why. I don't think it's a big deal though. The test tests the images 0023 migration, and we can remove that migration sometime after the rollout anyway (when handling issue #146). I just skipped that test for now. Just in case, I'll set up a Python 3 environment on our staging server and see if that encounters any issues with the new migrations.
Ok.

> 
> The second skip I just did was an unrelated test which has been unreliable on my Windows dev machine. Sometimes it fails, then the problem goes away later. Looks like the unreliability is passing onto Travis as well.

Great -- I have been seeing that fail intermittently also. I thought it was related to migrations. Glad to see it skipped.

**Comment by beijbom** [comment]
> As for testing on my end: no unit test failures, and the small-size regtest was able to complete successfully on the 4 example configs you gave in the help text: True local local, False local local, False local S3, False SQS S3.
> 
> Though, the 3rd regtest config (False local S3) did encounter a couple of intermittent errors. My first try got `ValueError: Not enough classes to do training (only 1)`, but eventually it tried training again and was successful. Second and third tries stopped short with an error like `OSError: File does not exist: media/backend_job_res/1592365319.3996203.json`. 4th time had no errors. That 'NOTE2' comment looks like you're aware of some intermittent issue with S3, but not sure if it's the same as the ones I got.

Yeah, the second and third tries are due to something in the timing being off between the local queue and the s3 storage. It is related to the NOTE2 that you commented on, although this is for the inverse setting of CELERY_EAGER.  I'll make a note to look into that later. 

> 
> Overall though, PR looks good to me. Particularly happy to see spacer can run without S3 storage or SQS.

Thanks! I'm very happy to have a bit more structure.

**Comment by beijbom** [comment]
@StephenChan : I merged this now. Let's continue conversation in base branch.

**Pull request merged**